File: election-backend/src/main/java/com/example/electionbackend/service/MetadataService.java
package com.example.electionbackend.service;

import com.example.electionbackend.interfaces.IMetadataService;
import com.example.electionbackend.model.Metadata;
import com.example.electionbackend.repository.interfaces.MetadataRepository;
import com.example.electionbackend.type.RegionType;
import org.springframework.stereotype.Service;

import java.util.List;

/**
 * Service class for managing Metadata entities.
 */
@Service
public class MetadataService implements IMetadataService {
    private final MetadataRepository metadataRepository;

    /**
     * Constructor for MetadataService.
     * @param metadataRepository - the metadata repository
     */
    public MetadataService(MetadataRepository metadataRepository) {
        this.metadataRepository = metadataRepository;
    }

    /**
     * Get Metadata by Region Type and ID
     * @param regionType - the type of the Region
     * @param regionId - the ID of the Region
     * @return - the list of Metadata associated with the Region Type and ID
     */
    @Override
    public List<Metadata> getByRegionTypeAndId(RegionType regionType, String regionId) {
        return metadataRepository.findByRegionTypeAndId(regionType, regionId);
    }

    /**
     * Get Metadata by Region Type
     * @param regionType - the type of the Region
     * @return - the list of Metadata associated with the Region Type
     */
    @Override
    public List<Metadata> getByRegionType(RegionType regionType) {
        return metadataRepository.findByRegionType(regionType);
    }
}

File: election-frontend/src/services/MetadataService.ts
import {IMetadataService} from "@/interfaces/IMetadataService";
import {IMetadata} from "@/interfaces/IMetadata";

/**
 * Service for fetching metadata from the backend API.
 */
export class MetadataService implements IMetadataService {
    private url: string = import.meta.env.VITE_API_URL as string;

    /**
     * Fetches metadata for a specific region type and ID from the backend API.
     * @param regionType - The type of the region.
     * @param regionId - The ID of the region.
     */
    public async getByRegionTypeAndId(regionType: string, regionId: string): Promise<IMetadata | null> {
        console.log(`${this.url}/metadata/region/${regionType}/${regionId}`); // TODO remove after testing
        const response: Response = await fetch(`${this.url}/metadata/region/${regionType}/${regionId}`, {
            method: 'GET',
            credentials: 'include',
            headers: {
                'Content-Type': 'application/json'
            }
        });
        if (!response.ok) {
            throw new Error('Failed to fetch metadata');
        }
        return await response.json();
    }

    /**
     * Fetches a list of metadata for a specific region type from the backend API.
     * @param regionType - The type of the region.
     */
    public async getByRegionType(regionType: string): Promise<IMetadata | null> {
        const response: Response = await fetch(`${this.url}/metadata/region/${regionType}`, {
            method: 'GET',
            credentials: 'include',
            headers: {
                'Content-Type': 'application/json'
            }
        });
        if (!response.ok) {
            throw new Error('Failed to fetch metadata');
        }
        return await response.json();
    }
}